pkg/runner/healthcheck_common.go

```diff
@@ -0,0 +1,227 @@
+type ErrgroupHealthcheckHelper struct {
+	toDo   []*toDoElement
+	report *api.HealthcheckReport
+}
+
+func (hh *ErrgroupHealthcheckHelper) Enlist(name string, c Checker, f Fixer) {
+	hh.toDo = append(hh.toDo, &toDoElement{name, c, f})
+}
+
+func (hh *ErrgroupHealthcheckHelper) RunChecks(ctx context.Context, fix bool) error {
+	eg, _ := errgroup.WithContext(ctx)
+
+	for _, li := range hh.toDo {
+		hcp := *li
+		eg.Go(func() error {
+			hcp := hcp
+			// Checker succeeds, already working.
+			if hcp.Checker() {
+				hh.report.Checks = append(hh.report.Checks, api.HealthcheckItem{
+					Name:    li.Name,
+					Status:  api.HealthcheckStatusOK,
+					Message: fmt.Sprintf("%s: OK", li.Name),
+				})
+				return nil
+			}
+			// Checker failed, Append the failure to the check report
+			hh.report.Checks = append(hh.report.Checks, api.HealthcheckItem{
+				Name:    li.Name,
+				Status:  api.HealthcheckStatusFailed,
+				Message: fmt.Sprintf("%s: FAILED. Fixing: %t", li.Name, fix),
+			})
+			// Attempt fix if fix is enabled.
+			fixhc := api.HealthcheckItem{Name: li.Name}
+			if fix {
+				err := li.Fixer()
+				if err != nil {
+					// Oh no! the fix failed.
+					fixhc.Status = api.HealthcheckStatusFailed
+					fixhc.Message = fmt.Sprintf("%s FAILED: %v", li.Name, err)
+				} else {
+					// Fix succeeded.
+					fixhc.Status = api.HealthcheckStatusOK
+					fixhc.Message = fmt.Sprintf("%s RECOVERED", li.Name)
+				}
+			} else {
+				// don't attempt to fix.
+				fixhc.Status = api.HealthcheckStatusOmitted
+				fixhc.Message = fmt.Sprintf("%s recovery not attempted.", li.Name)
+			}
+			// Fill the report with fix information.
+			hh.report.Fixes = append(hh.report.Fixes, fixhc)
+			return nil
+		})
+		eg.Wait() // TODO... Doing something wrong here. shouldn't have to be serial.
+	}
+	return eg.Wait()
+}
+
+// DefaultContainerChecker returns a Checker, a method which when executed will check for the
+// existance of the container. This should be considered a sensible default for checking whether
+// docker containers are started.
+func DefaultContainerChecker(ctx context.Context, log *zap.SugaredLogger, cli *client.Client, name string) Checker {
+	return func() bool {
+		ci, err := docker.CheckContainer(ctx, log, cli, name)
+		if err != nil || ci == nil {
+			return false
+		}
+		return ci.State.Running
+
+	}
+}
+
+// DefaultContainerFixer returns a Fixer, a method which when executed will ensure the named
+// container exists with some default paramaters which are appropriate for infra containers.
+// Unless containers require special consideration, this should be considered the sensible default
+// fixer for docker containers.
+func DefaultContainerFixer(ctx context.Context, log *zap.SugaredLogger, cli *client.Client, containerName string, imageName string, networkID string, portSpecs []string, pull bool, customHostConfig *container.HostConfig, cmds ...string) Fixer {
```

Comment:
This is different than the old `ensureInfraContainer` in a couple of ways.
1. It allows you to pass custom build options so it can be used with the sidecar.
2. the port mapping is explicit, rather than random, so containers have predictable port numbers in the default netns and can therefore be used by the local:exec runner.